tianshou/env/vecenv/asyncenv.py

```diff
@@ -0,0 +1,103 @@
+import gym
+import numpy as np
+from multiprocessing import connection
+from typing import List, Tuple, Union, Optional, Callable, Any
+
+from tianshou.env import SubprocVectorEnv
+
+
+class AsyncVectorEnv(SubprocVectorEnv):
+    """Vectorized asynchronous environment wrapper based on subprocess.
+    .. seealso::
+
+        Please refer to :class:`~tianshou.env.BaseVectorEnv` for more detailed
+        explanation.
+    """
+    def __init__(self, env_fns: List[Callable[[], gym.Env]],
+                 wait_num: Optional[int] = None) -> None:
+        """
+        :param wait_num: used in asynchronous simulation if the time cost of
+            ``env.step`` varies with time and synchronously waiting for all
+            environments to finish a step is time-wasting. In that case, we
+            can return when ``wait_num`` environments finish a step and keep
+            on simulation in these environments. If ``None``, asynchronous
+            simulation is disabled; else, ``1 <= wait_num <= env_num``.
+        """
+        super().__init__(env_fns)
+        self.wait_num = wait_num or len(env_fns)
```

Comment:
So if `wait_num` is None, this AsyncVenv is equal to the original `SubprocVenv`. Why not add this directly into the SubprocVenv?
By other words, the sync mode is a special case of async mode, thus we can unify the collector to use async mode for all venv instead of sync mode, and implement async in all venv.

Reply:
Maybe we can create another issue and pr

Reply:
Good point, and this can go further: why not just make it an ability of BaseVecEnv so that async simulation directly works for any vec envs? This seems worthwhile, but is a breaking change. Maybe we need to abstract Worker and change the interface of vec env.

Reply:
Having a unified VecEnv would be awesome and it may greatly ease the implementation of other classes/methods such as Collector.

Reply:
> In other words, the sync mode is a special case of async mode, thus we can unify the collector to use async mode for all venv instead of sync mode, and implement async in all venv.

That's what I was thinking x)

Reply:
@Trinkle23897 But I think adding this feature in a dedicated class as we did is a good start I think, since it may be break changing as said.